contracts/token/ERC721/ERC721Token.sol

```diff
@@ -1,45 +1,44 @@
 pragma solidity ^0.4.18;
 
 import "./ERC721.sol";
-import "../../math/SafeMath.sol";
+import "./DeprecatedERC721.sol";
+import "./ERC721BasicToken.sol";
+
 
 /**
- * @title ERC721Token
- * Generic implementation for the required functionality of the ERC721 standard
+ * @title Full ERC721 Token
+ * This implementation includes all the required and some optional functionality of the ERC721 standard
+ * Moreover, it includes approve all functionality using operatable terminology
+ * @dev see https://github.com/ethereum/eips/issues/721
  */
-contract ERC721Token is ERC721 {
-  using SafeMath for uint256;
-
-  // Total amount of tokens
-  uint256 private totalTokens;
+contract ERC721Token is ERC721, ERC721BasicToken {
+  // Token name
+  string internal name_;
 
-  // Mapping from token ID to owner
-  mapping (uint256 => address) private tokenOwner;
-
-  // Mapping from token ID to approved address
-  mapping (uint256 => address) private tokenApprovals;
+  // Token symbol
+  string internal symbol_;
 
   // Mapping from owner to list of owned token IDs
-  mapping (address => uint256[]) private ownedTokens;
+  mapping (address => uint256[]) internal ownedTokens;
 
   // Mapping from token ID to index of the owner tokens list
-  mapping(uint256 => uint256) private ownedTokensIndex;
+  mapping(uint256 => uint256) internal ownedTokensIndex;
 
-  /**
-  * @dev Guarantees msg.sender is owner of the given token
-  * @param _tokenId uint256 ID of the token to validate its ownership belongs to msg.sender
-  */
-  modifier onlyOwnerOf(uint256 _tokenId) {
-    require(ownerOf(_tokenId) == msg.sender);
-    _;
-  }
+  // Array with all token ids, used for enumeration
+  uint256[] internal allTokens;
+
+  // Mapping from token id to position in the allTokens array
+  mapping(uint256 => uint256) internal allTokensIndex;
```

Comment:
I may be missing something, but why is this mapping needed? As per the mint function, isn't the token's identifier the same thing as its index in the allTokens array?

`allTokensIndex[_tokenId] = allTokens.length;`

The only other place this gets read is during `burn`, where you could just use the `_tokenId` to infer location instead of this mapping.

Reply:
When a token is burnt, the one with the highest id is put in the place of the burnt one in the tokens array.

https://github.com/spalladino/zeppelin-solidity/blob/6fbe771c276f83e734256908877c8f9aaa2e1bfc/contracts/token/ERC721/ERC721Token.sol#L189

In this way the index ⇔ id equivalence is invalidated.

Reply:
> As per the mint function, isn't the token's identifier the same thing as its index in the allTokens array?

Yes, but only because `mint` generates autoincremental token IDs. I didn't want to rely on this for implementing the enumeration, as anyone should be able to override this behaviour if needed.